spec/index.bs

```diff
@@ -4183,41 +4189,71 @@ has a default layout created and used instead.
-        **Arguments:**
-            - {{GPUShaderStage}} |stage|
-            - {{GPUProgrammableStage}} |descriptor|
-            - {{GPUPipelineLayout}} |layout|
+        Each key-value pair must uniquely identify a [=pipeline-overridable=] constant
+        (by a string representing the numeric ID of the constant, if specified;
+        otherwise, by the constant's identifier name).
 
-        Return `true` if all of the following conditions are satisfied:
+        When the pipeline is executed, that constant will be set to the specified value
+        ([=converted to an IDL value|converted=] from `double` to
+            {{boolean}}, {{float}}, {{long}}, or {{unsigned long}},
+            depending on the [=wgsl-type=] of the [=wgsl-declaration=] in WGSL).
+</dl>
 
-            - The |descriptor|.{{GPUProgrammableStage/module}} is [=valid=] {{GPUShaderModule}}.
-            - The |descriptor|.{{GPUProgrammableStage/module}} contains
-                an entry point at |stage| named |descriptor|.{{GPUProgrammableStage/entryPoint}}.
-            - For each |binding| that is [=statically used=] by the shader entry point,
-                the [$validating shader binding$](|binding|, |layout|) returns `true`.
-            - For each texture sampling shader call that is [=statically used=] by the entry point:
-                1. Let |texture| be the {{GPUBindGroupLayoutEntry}} corresponding to the sampled texture in the call.
-                1. Let |sampler| be the {{GPUBindGroupLayoutEntry}} corresponding to the used sampler in the call.
-                1. One of the following conditions is `false`:
-                    - |texture|.{{GPUTextureBindingLayout/sampleType}} is {{GPUTextureSampleType/"unfilterable-float"}}:
-                    - |sampler|.{{GPUSamplerBindingLayout/type}} is {{GPUSamplerBindingType/"filtering"}}.
+<div algorithm>
+    <dfn abstract-op>validating GPUProgrammableStage</dfn>(stage, descriptor, layout)
 
+    **Arguments:**
+    - {{GPUShaderStage}} |stage|
+    - {{GPUProgrammableStage}} |descriptor|
+    - {{GPUPipelineLayout}} |layout|
+
+    Return `true` if all of the following conditions are met:
+
+    - |descriptor|.{{GPUProgrammableStage/module}} must be a [=valid=] {{GPUShaderModule}}.
+    - |descriptor|.{{GPUProgrammableStage/module}} must contain
+        an entry point, for shader stage |stage|,
+        named |descriptor|.{{GPUProgrammableStage/entryPoint}}.
+    - For each |binding| that is [=statically used=] by the shader entry point:
+        - [$validating shader binding$](|binding|, |layout|) must return `true`.
+    - For each texture sampling shader call that is [=statically used=] by the entry point:
+        1. Let |texture| be the {{GPUBindGroupLayoutEntry}} corresponding to the sampled texture in the call.
+        1. Let |sampler| be the {{GPUBindGroupLayoutEntry}} corresponding to the used sampler in the call.
+        1. One of the following conditions must be `false`:
```

Comment:
This seems like a confusing way to specify this rule? Seems to me like the following works just as well:

`If |sampler|.{{GPUSamplerBindingLayout/type}} is {{GPUSamplerBindingType/"filtering"}} then |texture|.{{GPUTextureBindingLayout/sampleType}} must not be {{GPUTextureSampleType/"unfilterable-float"}}.

Reply:
I didn't really change this (just changed the indentation and `is` -> `must be`) but I'll try to clean this up since I'm cleaning up other wording here too.

Reply:
Also I agree, your suggested wording is good.